Adding code changes for json form editor for managing spalp configurations

shashikant-walkunde-cg opened the pull request:
* Implemented json form builder and editor to manage app configurations using json schema structures in spalp app landing node. 

* Added sample configuration schema in example module to which is used for rendering config editor form on example app landing node edit

* JSON form editor is implemented using http://brutusin.org/json-forms/ library to generate configuration form. 

Please download the above package an place it inside drupal's libraries folder.

OR add following package dependency insider composer.json
```
"repositories": [

{
    "type": "package",
    "package": {
        "name": "brutusin/json-forms",
        "version": "0.0.0",
        "type": "drupal-library",
        "source": {
            "url": "https://github.com/brutusin/json-forms.git",
            "type": "git",
            "reference": "origin/master"
        }
    }
}

],
```

Then run `composer require brutusin/json-forms` to get this dependency inside drupal library.

* Additionally included configurations in applanding content type to include json configuration field to store edited configuration derived from json form. Attached config form widget to json configuration field in node form display.

shashikantwalkundecap committed:
Adding code changes for json form editor on spalp configuration editing using json schema from example module

shashikantwalkundecap committed:
Removed commented console logs

shashikantwalkundecap committed:
Fixed broken indents

malcomio commented on js/spalp-json-form-builder.js:
```diff
@@ -0,0 +1,50 @@
+      if (typeof settings.spalpJsonFormBuilder != 'undefined' && settings.spalpJsonFormBuilder != null) {
+        $.each(settings.spalpJsonFormBuilder, function (key, jsonFormElement) {
+          if (schemaDefined(jsonFormElement)) {
+            $("#" + jsonFormElement.identifier, context).once('spalp_json_form_builder').each(function () {
```
would be good to add some comments in here

malcomio commented on spalp_example/spalp_example.config.schema.json:
```diff
@@ -0,0 +1,72 @@
+            "firstParam": {
+              "type": "string"
+            },
+            "secorndParam": {
```
typo - `secornd`

malcomio commented:
please add the relevant lines to `composer.json` to this module for managing the dependency

malcomio committed:
Merge branch '8.x-1.x' into config-schema-editor

sharique commented on spalp_example/spalp_example.config.schema.json:
```diff
@@ -0,0 +1,72 @@
+          },
+          "required": [
+            "firstParam",
+            "secorndParam"
```
typo here again

malcomio commented on src/Service/SpalpConfig.php:
```diff
@@ -47,4 +47,26 @@ public function getAppIds() {
+   * @return array
+   *   Array representation of the configuration schema settings.
+   *
+   * @todo Merge this method with getConfigFromJson method present on pull request #7
```
#7 has been merged, so let's do this TODO

shashikant-walkunde-cg replied:
yes.. i see #7 merged now itself..

malcomio commented on config/install/core.entity_form_display.node.applanding.default.yml:
```diff
@@ -6,6 +6,7 @@ dependencies:
     - field.field.node.applanding.field_spalp_app_id
     - field.field.node.applanding.field_spalp_app_text
     - field.field.node.applanding.field_spalp_app_config
+    - field.field.node.applanding.field_spalp_config_json
```
should this replace the other two config fields, or do we need to apply the JSON work to the existing fields instead?

we definitely shouldn't end up with two ways to edit the same data

and I think we want to keep text and config as separate fields

shashikant-walkunde-cg replied:
This will replace other two configs.. I think if we are storing all configurations in node then have it as single field only.

malcomio replied:
in that case, let's delete the other two, and we'll need to update the `createNodes` function to match: https://github.com/Capgemini/spalp/blob/8.x-1.x/src/Service/Core.php#L60

I think @dakkusingh had an opinion that it's better to have them as two fields, but I don't remember why

malcomio replied:
keeping it as two fields would fit with the yml approach that @dakkusingh had been talking about, but I don't think we need to worry about it

let's go ahead with moving it into a single field

malcomio replied:
removing the extra fields is done in #15

malcomio commented on js/spalp-json-form-builder.js:
```diff
@@ -0,0 +1,50 @@
+        $.each(settings.spalpJsonFormBuilder, function (key, jsonFormElement) {
+          if (schemaDefined(jsonFormElement)) {
+            $("#" + jsonFormElement.identifier, context).once('spalp_json_form_builder').each(function () {
+              const brutsin_forms = brutusin["json-forms"];
```
please be consistent with the naming - is it `brutsin` or `brutusin`?

malcomio commented on js/spalp-json-form-builder.js:
```diff
@@ -0,0 +1,50 @@
+            $("#" + jsonFormElement.identifier, context).once('spalp_json_form_builder').each(function () {
+              const brutsin_forms = brutusin["json-forms"];
+              const container = document.getElementById(jsonFormElement.identifier);
+              const bf = brutsin_forms.create(jsonFormElement.schema);
```
`bf` isn't a very descriptive variable name

malcomio commented on js/spalp-json-form-builder.js:
```diff
@@ -0,0 +1,50 @@
+    return typeof (settings.schema) != 'undefined' && settings.schema != null
+        && typeof (settings.identifier) != 'undefined' && settings.identifier != null;
+  }
+})(jQuery, Drupal);
```
blank line at end of file please

malcomio commented on spalp_example/spalp_example.config.schema.json:
```diff
@@ -0,0 +1,72 @@
+    "config",
+    "app_text"
+  ]
+}
```
blank line at end please

malcomio commented on src/Plugin/Field/FieldWidget/JsonConfigFormWidget.php:
```diff
@@ -0,0 +1,114 @@
+    if (Unicode::strlen($element['value']['#value'])) {
+      Json::decode($element['value']['#value']);
+      if (json_last_error() !== JSON_ERROR_NONE) {
+        $form_state->setError($element['value'], t('!name must contain a valid data.', ['!name' => $element['#title']]));
```
confusing message - what is `a valid data`? i.e. how does the editor know what kind of data is valid?

malcomio commented on spalp_example/spalp_example.config.schema.json:
```diff
@@ -0,0 +1,72 @@
+    "app_text": {
+      "type": "object",
+      "properties": {
+        "firstField": {
```
ideally we'd have meaningful field names that match the existing config.json file
https://github.com/Capgemini/spalp/blob/8.x-1.x/spalp_example/spalp_example.config.json

malcomio commented on spalp_example/spalp_example.config.schema.json:
```diff
@@ -0,0 +1,72 @@
+        "nestedConfig": {
+          "type": "object",
+          "properties": {
+            "firstParam": {
```
let's have more meaningful names, as per https://github.com/Capgemini/spalp/blob/8.x-1.x/spalp_example/spalp_example.config.json

malcomio commented on src/Plugin/Field/FieldWidget/JsonConfigFormWidget.php:
```diff
@@ -0,0 +1,114 @@
+    $field_name = $this->fieldDefinition->getName();
+    $identifier = $field_name . '-' . $delta . '-jsonconfigform';
+    $selector = '[data-drupal-selector="' . Html::getId('edit-' . $field_name . '-' . $delta . '-value') . '"]';
+    $app_id = $form['field_spalp_app_id']['widget']['#default_value'][0];
```
this code is all quite dense - please add some blank lines and comments to make it more readable

malcomio commented on src/Plugin/Field/FieldWidget/JsonConfigFormWidget.php:
```diff
@@ -0,0 +1,114 @@
+   */
+  public static function validateJsonConfiguration(&$element, FormStateInterface $form_state, $form) {
+    if (Unicode::strlen($element['value']['#value'])) {
+      Json::decode($element['value']['#value']);
```
not obvious what's happening - let's add some comments

aaronferris commented on config/install/field.field.node.applanding.field_spalp_config_json.yml:
```diff
@@ -0,0 +1,18 @@
+label: 'App Configuration JSON'
+description: ''
+required: false
+translatable: false
```
Is this right?

aaronferris commented on src/Plugin/Field/FieldWidget/JsonConfigFormWidget.php:
```diff
@@ -0,0 +1,114 @@
+    }
+    $element['value'] = $element + [
+      '#type' => 'textarea',
+      '#suffix' => '<div id="' . $identifier . '"></div>',
```
Whats the purpose of this? Wouldnt it make sense for this element to be wrapped in this $identifier ID rather than have it as a suffix?

aaronferris commented on src/Service/SpalpConfig.php:
```diff
@@ -47,4 +47,26 @@ public function getAppIds() {
+    $filename = DRUPAL_ROOT . '/' . drupal_get_path('module', $module) . "/$module.config.schema.json";
+    if (file_exists($filename)) {
+      $string = file_get_contents($filename);
+      $json = json_decode($string);
```
We seem to be using Json::decode elsewhere in this PR, which is better to use?

dakkusingh replied:
I left same comment here
https://github.com/Capgemini/spalp/pull/14/files#r216594074

We should be using Json::decode
See: https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Component%21Serialization%21Json.php/function/Json%3A%3Adecode/8.2.x

aaronferris commented on src/Service/SpalpConfig.php:
```diff
@@ -47,4 +47,26 @@ public function getAppIds() {
+    // Get the JSON file for the module.
+    $filename = DRUPAL_ROOT . '/' . drupal_get_path('module', $module) . "/$module.config.schema.json";
+    if (file_exists($filename)) {
+      $string = file_get_contents($filename);
```
Is it possible for the file to exist but be empty? I assume not.

shashikantwalkundecap committed:
Updated getConfigFromJson method to also return schema config json files

malcomio commented on spalp.libraries.yml:
```diff
@@ -0,0 +1,18 @@
+json-form:
+  remote: https://github.com/brutusin/json-forms
+  version: VERSION
+  license:
```
looking at the example on https://www.drupal.org/project/drupal/issues/2276219, we need to link to the remote license as well

malcomio commented:
## To do before we merge
- [ ] add library to composer.json
- [x] remove extra fields
- [x] import data into new field - probably best to merge #13 first
- [x] hide JSON textarea when derived fields are added
- [x] JSON schema should match example config JSON
- [x] link to remote license

malcomio commented on js/spalp-json-form-builder.js:
```diff
@@ -0,0 +1,50 @@
+                "json_config_field": jsonFormElement.textarea
+              };
+              $("#" + jsonFormElement.identifier + " input").on('change', event_data, function (event) {
+                $(event.data.json_config_field).val(JSON.stringify(event.data.brutsin_form.getData()));
```
after this, let's hide the original textarea

shashikantwalkundecap committed:
Merge remote-tracking branch 'origin' into config-schema-editor

shashikantwalkundecap committed:
Added following change

1. Used spalp defined json to create schema for json form editor
2. Removed code references for spalp app text and app config individual fields
3. Merged latest 8.x.1.x into this branch to get #13 and used spalp_json_config field to store initial json
4. hide JSON textarea when derived fields are added
5. link to remote license

Pending :
1. add library to composer.json

shashikantwalkundecap committed:
Code cleanup removing unwanted method

passionatepoorna committed:
Issue #2987397 by Poornachandran: Multilingual

* Multilingual configuration for app landing pages
* Making the body field translatable
* Adding translation spalp config field
* Removing the spalp app config and app text fields along with their multilingual configs
